crates/gitbutler-repo/src/change_reference.rs

```diff
@@ -0,0 +1,213 @@
+) -> Result<ChangeReference> {
+    // The reference must be parseable as a remote reference
+    gitbutler_reference::RemoteRefname::from_str(&name)
+        .context("Failed to parse the provided reference")?;
+    let handle = VirtualBranchesHandle::new(ctx.project().gb_dir());
+    // The branch must exist
+    let mut vbranch = handle.get_branch(branch_id)?;
+
+    // Enusre that the commit acutally exists
+    let new_commit =
+        commit_by_branch_id_and_change_id(ctx, &vbranch, &handle, new_change_id.clone())
+            .context(anyhow!("Commit for change_id {} not found", new_change_id))?;
+
+    // Fail early if the commit is not valid
+    validate_commit(&vbranch, new_commit.id(), ctx, &handle)?;
+
+    let reference = vbranch
+        .references
+        .iter_mut()
+        .find(|r| r.name == name)
+        .ok_or(anyhow!(
+            "Reference {} not found for branch {}",
+            name,
+            branch_id
+        ))?;
+    reference.change_id = new_change_id;
+    let new_reference = reference.clone();
+    handle.set_branch(vbranch)?;
+    Ok(new_reference)
+}
+
+/// Pushes a gitbutler branch reference to the remote repository.
+pub fn push_change_reference(
+    ctx: &CommandContext,
+    branch_id: BranchId,
+    name: ReferenceName,
+    with_force: bool,
+    credentials: &Helper,
+) -> Result<()> {
+    let handle = VirtualBranchesHandle::new(ctx.project().gb_dir());
+    let vbranch = handle.get_branch(branch_id)?;
+    let reference = vbranch
+        .references
+        .iter()
+        .find(|r| r.name == name)
+        .ok_or_else(|| anyhow!("Reference {} not found", name))?;
+    let upstream_refname = gitbutler_reference::RemoteRefname::from_str(&reference.name)
+        .context("Failed to parse the provided reference")?;
+    let commit =
+        commit_by_branch_id_and_change_id(ctx, &vbranch, &handle, reference.change_id.clone())?;
+    ctx.push(
+        &commit.id(),
+        &upstream_refname,
+        with_force,
+        credentials,
+        None,
+        Some(Some(branch_id)),
+    )
+}
+
+/// Given a branch id and a change id, returns the commit associated with the change id.
+// TODO: We need a more efficient way of getting a commit by change id.
+fn commit_by_branch_id_and_change_id<'a>(
+    ctx: &'a CommandContext,
+    vbranch: &Branch,
+    handle: &VirtualBranchesHandle,
+    change_id: String,
+) -> Result<git2::Commit<'a>> {
+    let target = handle.get_default_target()?;
+    let branch_commits = ctx
+        .log(vbranch.head, LogUntil::Commit(target.sha))?
+        .iter()
+        .map(|c| c.id())
+        .collect_vec();
+    // Find the commit with the change id
+    let commit = branch_commits
+        .iter()
+        .find(|c| {
+            let commit = ctx.repository().find_commit(**c).expect("Commit not found");
+            commit.change_id() == Some(change_id.clone())
+        })
+        .map(|c| ctx.repository().find_commit(*c).expect("Commit not found"))
```

Comment:
Let's avoid panics, the UI will not respond kindly.

```suggestion
        .and_then(|c| ctx.repository().find_commit(*c).ok())
```